MCSync/Syncer.cs

```diff
@@ -1,326 +1,359 @@
-                    string.IsNullOrEmpty(configValues.GetOrDefault("modsDirectory", string.Empty).Cast<string>()) || string.IsNullOrEmpty(configValues.GetOrDefault("resourcePackDirectory", "").Cast<string>()) || string.IsNullOrEmpty(configValues.GetOrDefault("configFilesDirectory", "").Cast<string>()) || !Enum.TryParse(configValues["mode"].Cast<string>(), true, out SyncMode configSyncMode) ||
-                    (configSyncMode == SyncMode.Client && (!configValues.ContainsKey("oResourcesDirectory") || string.IsNullOrEmpty(configValues["oResourcesDirectory"].Cast<string>()))))
-                {
-                    Task.Errors.Add("Configuration file not setup");
-                    SyncComplete?.Invoke(this, new EventArgs());
-                    return;
-                }
-
-                string modsDirectory = configValues["modsDirectory"].Cast<string>();
-                string resourcePackDirectory = configValues["resourcePackDirectory"].Cast<string>();
-                string configFilesDirectory = configValues["configFilesDirectory"].Cast<string>();
-                string oResourcesDirectory = configValues["oResourcesDirectory"].Cast<string>();
-
-                string[] clientSideWhiteListMods = UserPreferences.Get().Sections.GetOrDefault("mcsync", new Dictionary<string, object>()).GetOrDefault("mods_whitelist")?.Cast<string[]>() ?? new string[0];
-                string[] clientSideWhiteListResourcePacks = UserPreferences.Get().Sections.GetOrDefault("mcsync", new Dictionary<string, object>()).GetOrDefault("resourcepacks_whitelist")?.Cast<string[]>() ?? new string[0];
-
+        }
+
+        public async void BeginSync()
+        {
+            await System.Threading.Tasks.Task.Run(DoSync);
+        }
+
+        public static bool IsDownloadTypeValid(SyncMode mode, MCSyncFile.DownloadTypes downloadTypes)
+        {
+            if(downloadTypes == MCSyncFile.DownloadTypes.Client && mode == SyncMode.Client)
+            {
+                return true;
+            }
+            else if (downloadTypes == MCSyncFile.DownloadTypes.Server && mode == SyncMode.Server)
+            {
+                return true;
+            }
+            else if (downloadTypes == MCSyncFile.DownloadTypes.Common)
+            {
+                return true;
+            }
+            return false;
+        }
+
+        private async System.Threading.Tasks.Task DoSync()
+        {
+            try
+            {
+                Task.Informations.Clear();
+                Task.Errors.Clear();
+
+                // Load config
+                Dictionary<string, object> configValues = UserPreferences.Get().Sections.GetOrSetDefault("mcsync", new Dictionary<string, object>());
+
+                if (!configValues.ContainsKey("modsDirectory") || !configValues.ContainsKey("resourcePackDirectory") || !configValues.ContainsKey("configFilesDirectory") || !configValues.ContainsKey("mode") ||

```

Comment:
Need to add the `animationDirectory` into this `if` statement, otherwise we will get an null reference error if the user fails to set up the directory.